src/fundus/utils/more_async.py

```diff
@@ -0,0 +1,59 @@
+import asyncio
+from typing import AsyncIterator, Iterable, TypeVar, Union, overload
+
+_T = TypeVar("_T")
+_VT = TypeVar("_VT")
+
+
+class _Sentinel:
+    pass
+
+
+__sentinel = _Sentinel()
+
+
+@overload
+async def async_next(iterator: AsyncIterator[_T]) -> _T:
+    ...
+
+
+@overload
+async def async_next(iterator: AsyncIterator[_T], default: Union[_VT, _Sentinel]) -> Union[_T, _VT]:
+    ...
+
+
+async def async_next(iterator: AsyncIterator[_T], default: Union[_VT, _Sentinel] = __sentinel) -> Union[_T, _VT]:
+    task = iterator.__anext__()
+    try:
+        return await task
+    except StopAsyncIteration:
+        if not isinstance(default, _Sentinel):
+            return default
+        else:
+            raise StopAsyncIteration
+
+
+async def batched_async_interleave(*generators: AsyncIterator[_T]) -> AsyncIterator[Iterable[_T]]:
```

Comment:
Going with the `more_itertools` terminology this function does not fit a batched version of [`interleave`](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.interleave_longest) but [`interleave_longest`](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.interleave_longest). Maybe renaming it to `batched_interleave_longest` and linking the non-async variant would help to understand this function immediately.

I don't really understand the implementation of this function. But I haven't looked really hard into it. Nevertheless just from the output if it almost behaves like the more_itertools `inverleave_longest` we could simplify the function with the `aioitertools` library that implements itertools functions for async.

```python
async def batched_interleave_longest(*async_iterables: AsyncIterator[_T]) -> AsyncIterator[Iterator[_T]]:
    batches = aioitertools.zip_longest(*async_iterables, fillvalue=__sentinel)
    async for batch in batches:
        yield (x for x in batch if x is not __sentinel)
```

Full example with working code: 
[async_interleave.zip](https://github.com/flairNLP/fundus/files/11923566/async_interleave.zip)

Reply:
I agree that the naming is a bit off.
Regarding the implementation: This works like `zip_longest` but instead of yielding fill values, this skips empty iterables.

Your proposed function looks much cleaner :) I would do some testing and replace mine with this.
Thanks for proposing.

Reply:
We'll leave this to another PR.